### PR TITLE
EOS-17303: S3faulttolerence - Object Del (single object) API support

### PR DESCRIPTION
**Pull request by DPG17**
Note: Existing UT and ST are successful.

Signed-off-by: Dattaprasad <dattaprasad.govekar@seagate.com>

### CODE REVIEW

File: s3config.release.yaml.sample [hunk 1]
```diff
@@ -70,6 +70,7 @@ S3_SERVER_CONFIG:                                       # Section for S3 Server
    S3_SERVER_MOTR_ETIMEDOUT_MAX_THRESHOLD: 5            # Number of ETIMEDOUT errors per monitoring window before s3server restart
    S3_SERVER_MOTR_ETIMEDOUT_WINDOW_SEC: 60              # Monitoring window for motr ETIMEDOUT errors in seconds
    S3_SERVER_ENABLE_ADDB_DUMP: true                     # If set to true, then addb dump will be collected
+   S3_MAX_EXTENDED_OBJECTS_IN_FAULT_MODE: 0		# Maximum number of allowed recovery by S3 when fault happens and Motr error out with specific error code. When 0, feature is disabled. The maximum value is 10.
```
Comment (knrajnambiar76) [inline s3config.release.yaml.sample:73]: maximum we can reduce to 5. 10 is high

File: server/s3_delete_object_action.cc [hunk 2]
```diff
@@ -273,15 +333,42 @@ void S3DeleteObjectAction::delete_object() {
...
-    motr_writer = motr_writer_factory->create_motr_writer(
-        request, object_metadata->get_oid());
+  unsigned int object_count = del_object_oids.size();
+  if (object_count > 0) {
+    // process to delete object
+    std::pair<struct m0_uint128, int>& obj_layout_pair = del_object_oids.back();
+    if (!motr_writer) {
+      motr_writer = motr_writer_factory->create_motr_writer(
+          request, obj_layout_pair.first);
+    } else {
```
Comment (knrajnambiar76) [inline server/s3_delete_object_action.cc:343]: is this else needed ?

File: s3config.release.yaml.sample (hunk 1 above)
Comment (DPG17) [reply s3config.release.yaml.sample]: This is configurable. we can always set to any value. I can set it to 5 on your request, but I don't know why only 5, why not 8 or 10 itself.

File: server/s3_delete_object_action.cc (hunk 2 above)
Comment (DPG17) [reply server/s3_delete_object_action.cc:343]: yes, else part is required to set correct oid to be deleted. Now, we have responsibility to delete multiple objects.

File: s3config.release.yaml.sample (hunk 1 above)
Comment (knrajnambiar76) [reply s3config.release.yaml.sample]: In calls we said that we will be supporting one or twice failure, 10 is pretty big. Have it at lower value

File: server/s3_delete_object_action.cc (hunk 2 above)
Comment (knrajnambiar76) [reply server/s3_delete_object_action.cc:343]: But if motr_writer is not set then create_motr_writer() instantiate it after that setting of OID needed right

**Review by knrajnambiar76 — APPROVED** [review]

File: s3config.release.yaml.sample (hunk 1 above)
Comment (DPG17) [reply s3config.release.yaml.sample]: By default, value is 0 to make feature disabled. But, I've added in the comment of this property to set max to 5.

File: server/s3_delete_object_action.cc [hunk 3]
```diff
@@ -130,8 +130,12 @@ void S3DeleteObjectAction::add_object_oid_to_probable_dead_oid_list() {
 
   // prepending a char depending on the size of the object (size based bucketing
   // of object)
-  S3CommonUtilities::size_based_bucketing_of_objects(
-      oid_str, object_metadata->get_content_length());
+  size_t object_size = object_metadata->get_content_length();
+  if (object_metadata->is_object_extended()) {
```
Comment (amitwac2608) [inline server/s3_delete_object_action.cc:134]: should we add else block and read the content length there, instead of doing that in above line ?

**Review by amitwac2608 — COMMENTED** [review]
looks good.

File: server/s3_delete_object_action.cc (hunk 3 above)
Comment (DPG17) [reply server/s3_delete_object_action.cc:134]: well, no need of else block. The logic is first get size of object using get_content_length(). This should be sufficient for 95% of time, and we don't have to execute code within 'if' block. Only when object is extended, which is rare, we have to read object size using 'get_primary_obj_size()'. Moreover, the code also looks readable.

**Review by amitwac2608 — APPROVED** [review]

File: server/s3_delete_object_action.cc (hunk 3 above)
Comment (DPG17) [reply server/s3_delete_object_action.cc:134]: Actually, there is no much difference in the way you suggested here. So, any way can be done. My suggestion is let it be as is, .

**Comment by nileshgovande** [comment]
Please run jenkins job

**Pull request merged**